src/data/datasets/base.py

```diff
@@ -0,0 +1,147 @@
+from pathlib import Path
+from abc import ABC, abstractmethod
+from typing import Union, Optional
+
+import cv2
+import numpy as np
+import pandas as pd
+from albumentations import BasicTransform
+from albumentations.core.composition import BaseCompose
+from torch.utils.data import Dataset
+
+
+class ImageDataset(Dataset, ABC):
+    """ An abstract class for image dataset """
+
+    def __init__(self,
+                 data_folder: str,
+                 csv_path: str,
+                 transform: Optional[Union[BasicTransform, BaseCompose]],
+                 augment: Optional[Union[BasicTransform, BaseCompose]] = None,
+                 input_dtype: str = 'float32',
+                 input_column: str = 'image_path',
+                 grayscale: bool = False,
+                 test_mode: bool = False,
+                 transform_targets: dict = {'input': 'image'}):
+        """
+        Args:
+            data_folder: Directory with all the images.
+            csv_path: Path to the csv file with path to images and annotations.
+                Path to images must be under column `image_path` and annotations must be under `label` column.
+            transform: Transform to be applied on a sample. This should have the
+                interface of transforms in `albumentations` library.
+            augment: Optional augment to be applied on a sample.
+                This should have the interface of transforms in `albumentations` library.
+            input_dtype: Data type of of the torch tensors related to the image.
+            input_column: Name of the column that contains paths to images.
+            grayscale: If True, image will be read as grayscale otherwise as RGB.
+            test_mode: If True, only image without labels will be returned.
+            transform_target: Transformations target for `albumentations` library.
+        """
+        self.__test_mode = test_mode
+        self.__transform = transform
+        self.__augment = augment
+        self.__input_dtype = input_dtype
+        self.__input_column = input_column
+        self.__grayscale = grayscale
+        self.__transform_targets = transform_targets
+
+        self.__data_folder = Path(data_folder)
+        self.__csv = pd.read_csv(self.data_folder / csv_path)
```

Comment:
Move to child classes, put `dtype` for each known column.  
`self.__csv = pd.read_csv(self.data_folder / csv_path, dtype={'input_column': str, 'target_column': int or str})`